File: ROS2-basic/chapter5/chapter5_ws/src/demo_python_tf/demo_python_tf/tf_listener.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformListener, Buffer
from tf_transformations import quaternion_from_euler # 欧拉角转四元数函数
from tf_transformations import euler_from_quaternion

class TFListener(Node):

    def __init__(self):
     super().__init__('tf_listener')
     self.tf_buffer = Buffer()
     self.tf_listener = TransformListener(self.tf_buffer, self)
     self.timer_ = self.create_timer(1.0, self.timer_callback)
     self.get_logger().info('TFListener node has been started')

     def timer_callback(self):
        try:
            trans = self.tf_buffer.lookup_transform('base_link', 'bottle_link', rclpy.time.Time())
            trans_euler = euler_from_quaternion([trans.transform.rotation.x, trans.transform.rotation.y, trans.transform.rotation.z, trans.transform.rotation.w])
            self.get_logger().info('Euler angles: {}'.format(trans_euler))
            self.get_logger().info('Translation: {}'.format(trans.transform.translation))
            self.get_logger().info('Rotation: {}'.format(trans.transform.rotation))
        except Exception as e:
            self.get_logger().error('Error: {}'.format(e))

def main():
    rclpy.init()
    tf_listener = TFListener()
    rclpy.spin(tf_listener)
    rclpy.shutown()

File: ROS2-basic/chapter5/chapter5_ws/src/demo_cpp_tf/src/tf_listener.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/LinearMath/Quaternion.hpp"
#include "tf2/utils.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"
#include <memory>
#include <chrono>
#include <rclcpp/duration.hpp>
#include <tf2/utils.hpp>


class TFListener : public rclcpp::Node
{
	public:
		TFListener() : Node("tf_listener")
		{
			buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
			listener_ = std::make_shared<tf2_ros::TransformListener>(*buffer_);
			timer_ = this->create_wall_timer(std::chrono::seconds(5), std::bind(&TFListener::getTransform, this));
		}
		void getTransform()
		{
			try {
				const auto transform = buffer_->lookupTransform("base_link", "target_point",
									        this->get_clock()->now(),
										rclcpp::Duration::from_seconds(1.0f));
				const auto &translation = transform.transform.translation;
				const auto &rotation = transform.transform.rotation;
				double roll, pitch, yaw;
				tf2::getEulerYPR(rotation, yaw, pitch, roll);
				RCLCPP_INFO(this->get_logger(), "Translation: [%f, %f, %f]", translation.x, translation.y, translation.z);
				RCLCPP_INFO(this->get_logger(), "Rotation: [%f, %f, %f, %f]", rotation.x, rotation.y, rotation.z, rotation.w);
				RCLCPP_INFO(this->get_logger(), "Euler angles: [%f, %f, %f]", roll, pitch, yaw);
			}
			catch (tf2::TransformException &ex) {
				RCLCPP_WARN(this->get_logger(), "exception: %s", ex.what());
			}
		}

	private:
		rclcpp::TimerBase::SharedPtr timer_;
		std::shared_ptr<tf2_ros::TransformListener> listener_;
		std::shared_ptr<tf2_ros::Buffer> buffer_;

};

int main(int argc, char **argv)
{
	rclcpp::init(argc, argv);
	rclcpp::spin(std::make_shared<TFListener>());
	rclcpp::shutdown();
	return 0;
}

